File: leetcode/medium/238_product_except_self/product_except_self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [1] * len(nums)
        right = [1] * len(nums)
        for i in range(1, len(nums)):
            left[i] = left[i - 1] * nums[i - 1]
        for i in reversed(range(len(nums) - 1)):
            right[i] = right[i + 1] * nums[i + 1]
        result = []
        for i in range(len(nums)):
            result.append(left[i] * right[i])
        return result


print(Solution().productExceptSelf([1, 2, 3, 4]))

File: leetcode/medium/238_product_except_self/product_except_self.cpp
#include <vector>

class Solution {
public:
  std::vector<int> productExceptSelf(std::vector<int> &nums) {
    std::vector<int> result(nums.size(), 1);
    int left = 1;
    int right = 1;
    for (int i = 0; i < nums.size(); i++) {
      result[i] *= left;
      result[nums.size() - 1 - i] *= right;
      left *= nums[i];
      right *= nums[nums.size() - 1 - i];
    }
    return result;
  }
};
